doc/monads/laws.lean

```diff
@@ -0,0 +1,319 @@
+Transformers](transformers.lean.md).
+
+So far, you've learned the concrete details you need in order to _use_ monads in your Lean programs.
+But there's still one more important concept you need if you want to _create_ new functors,
+applicatives and monads. Namely, the notion of _structural "laws"_ -- rules that these type
+classes should follow in order to meet other programmers' expectations about your code.
+
+## Life without Laws
+
+Remember Lean represents each of these abstract classes by a type class. Each of these type classes
+has one or two main functions. So, as long as you implement those functions and it type checks, you
+have a new functor, applicative, or monad, right?
+
+Well not quite. Yes, your program will compile and you'll be able to use the instances. But this
+doesn't mean your instances follow the mathematical constructs. If they don't, your instances won't
+fulfill other programmers' expectations. Each type class has its own "laws". For instance, suppose
+you have the following Point Functor:
+-/
+structure Point (α : Type) where
+  x : α
+  y : α
+  deriving Repr, BEq
+
+def Point.map (f : α → β) (s : Point α) : Point β :=
+  { x := f s.y,  -- an example of something weird
+    y := f s.x }
+
+instance : Functor Point where
+  map := Point.map
+
+#eval (·+2) <$> (Point.mk 1 2) -- { x := 4, y := 3 }
+
+/-!
+This Point does something weird, when you `map` it because it transposes the `x` and `y` coordinates
+which is not what other people would expect from a `map` function.  In fact, it breaks the rules
+as you will see below.
+
+## What are the Functor laws?
+
+Functors have two laws: the _identity_ law, and the _composition_ law. These laws express behaviors that
+your functor instances should follow. If they don't, other programmers will be very confused at the
+effect your instances have on their program. Many structures have similar laws, including monads.
+
+The identity law says that if you "map" the identity function (`id`) over your functor, the
+resulting functor should be the same. A succinct way of showing this on a `List` functor is:
+
+-/
+def list1 := [1,2,3]
+
+#eval id <$> list1 == list1 -- true
+/-!
+
+Now let's try the same test on the `Point` functor:
+-/
+
+def p1 : Point Nat := (Point.mk 1 2)
+
+#eval id <$> p1 == p1 -- false
+
+/-!
+Oh, and look while the List is behaving well, the `Point` functor fails this identity test.
+
+The _composition_ law says that if you "map" two functions in succession over a functor, this
+should be the same as "composing" the functions and simply mapping that one super-function over the
+functor.  In Lean you can compose two functions using `Function.comp f g` (or the syntax `f ∘ g`,
+which you can type in VS code using `\o `) and you will get the same results from both of these
+showing that the composition law holds for `List Nat`:
+
+-/
+def double (x : Nat) := x + x
+def square (x : Nat) := x * x
+
+#eval (double <$> (square <$> list1)) -- [2, 8, 18]
```

Comment:
You don't need the outer parentheses here.
```
#eval double <$> (square <$> list1)
```

Reply:
fixed in https://github.com/leanprover/lean4/pull/1505/commits/8dd24bc9dae40eb19b4625bf0b94cfb3b15a2f07